### PR TITLE
Level override check is case-insensitive.

### LINKED ISSUE #1944: Only case sensitive check for minimumLevel override
**Description**
Serilog currently only supports case sensitive matches when overriding the log level for a specific namespace, when for example using Serilog with Serilog.Settings.Configuration and setting env variables for a docker image the Namespaces can't be all upper case.

**Reproduction**
Setting up a test project with Serilog and Serilog.Settings.Configuration installed, configure Serilog for example like this:

``` C#
host.UseSerilog((context, configuration) =>
        {
            configuration
                .WriteTo.Console(new JsonFormatter())
                .ReadFrom.Configuration(context.Configuration);
        })
```

add a appsettings file with the following configuration:

``` json
"Serilog": {
    "Using":  [ "Serilog.Sinks.Console" ],
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        "Test.Nmspc": "Information"
      }
    },
    "WriteTo": [
      { "Name": "Console" }
    ],
    "Enrich": [
      "FromLogContext"
    ],
    "Properties": {
      "Application": "Org.Eclipse.TractusX.Portal.Backend.Administration.Service"
    }
  },
```

Then execute the application with the following env var: SERILOG__MINIMUMLEVEL__OVERRIDE__TEST.NMSPC=Debug

The minimum logLevel for Loggers within that namespace will be Information

If needed I can provide an example repo.

**Expected behavior**
The minimum log level for logger instances creates within the Test.Nmspc Namespace should be Debug.

**Relevant package, tooling and runtime versions**
- Serilog 3.0.1
- Serilog.Settings.Configuration 3.4.0
- aspnet core 6 & 7 (both tested)
- docker image with environment variables (tested locally with a secret file as well on a win 10 machine)

**Additional context**
I've already provided a PR with a change for the case insensitive check.

### PR DESCRIPTION
**Pull request by tillig**
Fix #1944.

Changes `LevelOverrideMap.GetEffectiveLevel()` to use a case-insensitive comparison for context names, correcting some challenges with setting overrides through mechanisms like environment variables where ALL_CAPS may be used.

### CODE REVIEW

**Comment by nblumhardt** [comment]
@serilog/maintainers I'm on board with this one but suspect it needs another set of eyes on it, if just to spot other downstream places this could break. I'll merge in a week otherwise :-)

Thanks @tillig!

File: src/Serilog/Core/LevelOverrideMap.cs
Comment (bartelink) [inline src/Serilog/Core/LevelOverrideMap.cs:72]: Surely the second part of this condition is more selective and cheaper to compute ?
i.e.
            var loLen = levelOverride.Context.Length;
            if ((context.Length == loLen || (context.Length > loLen && context[loLen] == '.'))
                && context.StartsWith(levelOverride.Context, StringComparison.OrdinalIgnoreCase))
                
(my C# is too rusty to convert that to an `is` expression here!)

Comment (nblumhardt) [reply src/Serilog/Core/LevelOverrideMap.cs:72]: Oh, yes, good call @bartelink. The clause would need a slight modification:

```csharp
            if (context.Length >= levelOverride.Context.Length &&
                (context.Length == levelOverride.Context.Length || context[levelOverride.Context.Length] == '.') &&
                context.StartsWith(levelOverride.Context, StringComparison.OrdinalIgnoreCase))
```

@tillig I wonder if this would be enough to claw back a few % of the performance regression related to the case-sensitivity check? I'll dig in and take a look when I have a chance, if you don't beat me to it :-)

Comment (tillig) [reply src/Serilog/Core/LevelOverrideMap.cs:72]: I'm out of town for a week or so, so it'll be a bit before I'm back here. Happy to make the change, just will be a bit.

Comment (tillig) [reply src/Serilog/Core/LevelOverrideMap.cs]: I'm back and got this. Sorry for the delay.

Comment (tillig) [reply src/Serilog/Core/LevelOverrideMap.cs]: Hmmm. It appears swapping those has started causing tests to fail. I'll have to look into it.

Comment (tillig) [reply src/Serilog/Core/LevelOverrideMap.cs]: OK, I'm dumb, I missed the thing @nblumhardt mentioned.

I split the `==` and `>` so in the full match scenario it should be just slightly faster since it won't have to check for equality again. I'm sure it's on the order of microseconds, but 🤷 .

**Review by nblumhardt — APPROVED** [review]

**Comment by nblumhardt** [comment]
Thank you @tillig! 👍

File: src/Serilog/Core/LevelOverrideMap.cs
Comment (bartelink) [inline src/Serilog/Core/LevelOverrideMap.cs:73]: Sorry for the necroposting on something that's already overkill, but I will forever be pondering whether using a String.Equals into the same length case will be cheaper, or whether `StartsWith` internally special cases this...

Comment (khellang) [reply src/Serilog/Core/LevelOverrideMap.cs:73]: This is what the current `StartsWith` implementation does (when passing `OrdinalIgnoreCase`):

```csharp
case StringComparison.OrdinalIgnoreCase:
    if (this.Length < value.Length)
    {
        return false;
    }
    return Ordinal.EqualsIgnoreCase(ref this.GetRawStringData(), ref value.GetRawStringData(), value.Length);
```

And this is (effectively, I inlined a method call for brevity) what `Equals` does:

```csharp
case StringComparison.OrdinalIgnoreCase:
    if (a.Length != b.Length)
    {
        return false;
    }

    return Ordinal.EqualsIgnoreCase(ref a.GetRawStringData(), ref b.GetRawStringData(), b.Length);
```

In the case where the length is the same, they both call into the same machinery with the same length, so I think they should be equal here.

Comment (bartelink) [reply src/Serilog/Core/LevelOverrideMap.cs:73]: Thanks Kristian, this has made my day (especially for saving Travis the lookup from an annoying drive by commenter!)

**Pull request merged**
